### PR TITLE
Add `DustCollector`

### PR DESCRIPTION
**Pull request by AurevoirXavier**
`DustCollector` is a collector to manage the `DustLost` and `KilledAccount` when there're more than one currency on the chain.

Avoid killing the account,  if this account hold other currencies.

A pretty simple test case: `cargo t -p pallet-balances dust_collector_should_work`

---

If it's ok, I'll fix the remain tests and add some doc.

- [ ] test
- [ ] doc

### CODE REVIEW

File: frame/balances/src/lib.rs [hunk 1]
```diff
@@ -1351,3 +1386,23 @@ impl<T: Trait<I>, I: Instance> IsDeadAccount<T::AccountId> for Module<T, I> wher
...
+
+impl<T: Trait<I>, I: Instance> DustCollector<T::AccountId> for Module<T, I> {
+	fn is_collectable(who: &T::AccountId) -> bool {
+		let total = Self::total_balance(who);
+
+		if total < T::ExistentialDeposit::get() || total.is_zero() {
+			true
+		} else {
+			false
+		}
```
Comment (xlc) [inline frame/balances/src/lib.rs:1398]:
```suggestion
		total < T::ExistentialDeposit::get() || total.is_zero()
```

File: frame/support/src/traits.rs [hunk 2]
```diff
@@ -1664,6 +1664,28 @@ pub trait Instance: 'static {
     const PREFIX: &'static str ;
 }
 
+pub trait DustCollector<AccountId> {
```
Comment (kianenigma) [inline frame/support/src/traits.rs:1667]: This definitely needs to be clearly documented (and properly formatted).

File: frame/support/src/traits.rs [hunk 3]
```diff
@@ -1664,6 +1664,28 @@ pub trait Instance: 'static {
...
+	fn is_collectable(who: &AccountId) -> bool;
```
Comment (kianenigma) [inline frame/support/src/traits.rs:1668]: Am I wrong to think `is_dust()` is a better name for this? collecting doesn't quite imply that it is being killed.

**Review by kianenigma — COMMENTED** [review]
How does this fit with the ref-count that each account has? If we have multiple balance modules, to prevent one from killing the account if the account has tokens in another one, can't the ref-count be used somehow?

**Comment by kianenigma** [comment]
finishing (specially) the docs and tests will invite more people to try and review. @AurevoirXavier

**Comment by AurevoirXavier** [comment]
> How does this fit with the ref-count that each account has? If we have multiple balance modules, to prevent one from killing the account if the account has tokens in another one, can't the ref-count be used somehow?

I didn't find any usage of `in_ref/dec_ref`, except the `update_locks`. I think it's OK.

**Comment by kianenigma** [comment]
> Avoid killing the account, if this account hold other currencies.

Just saying that if you want this, then each balance pallet can increment ref and the account will not be killed, even if one of its currencies goes below zero.

**Comment by kianenigma** [comment]
I have been informed that we have also concluded that having such a functionality is useful, but as far as I know the proposed implementation is different from what you are doing here. I think we need to leave this stale/iced for a bit longer.

**Comment by AurevoirXavier** [comment]
> I have been informed that we have also concluded that having such a functionality is useful, but as far as I know the proposed implementation is different from what you are doing here. I think we need to leave this stale/iced for a bit longer.

Awesome. My implementation is a little bit ugly just a workaround for my chain and I've no time working on this now.

I expect this so much.

**Comment by kianenigma** [comment]
Can follow here. https://github.com/paritytech/substrate/pull/7152

**Pull request closed without merging**